### PR TITLE
[Impacting] Enable gzip compression by default

### PR DESCRIPTION
**Pull request by mougams**
Enable gzip compression on downstream and upstream. This will done by default, and no option introduced to disable it.

### CODE REVIEW

**Review by pierrecdn — APPROVED** [review]
Looks good to me.

File: haproxy/state/from_ha.go [hunk 1]
```diff
@@ -51,30 +51,44 @@ func FromHAProxy(ha HAProxyRead) (State, error) {
 		if err != nil {
 			return state, err
 		}
-		if len(logTargets) > 1 {
-			return state, fmt.Errorf("expected at most 1 filter for frontend %s, got %d", f.Name, len(filters))
+		if len(logTargets) >= 2 {
```
Comment (dclaisse) [inline haproxy/state/from_ha.go:54]: No change (">1" == ">=2")? Also, why is logTargets reused here? Looks like there was a typo initially, we should check length of filters here.

File: haproxy/state/from_ha.go [hunk 2]
```diff
@@ -51,30 +51,44 @@ func FromHAProxy(ha HAProxyRead) (State, error) {
...
+			return state, fmt.Errorf("expected at most 2 filters for frontend %s, got %d", f.Name, len(filters))
```
Comment (dclaisse) [inline haproxy/state/from_ha.go:55]: Message is not consistent with condition above.

File: haproxy/state/from_ha.go [hunk 3]
```diff
@@ -51,30 +51,44 @@ func FromHAProxy(ha HAProxyRead) (State, error) {
...
-		var filter *FrontendFilter
-		if len(filters) == 1 {
-			filter = &FrontendFilter{
-				Filter: filters[0],
+		var filter_spoe *FrontendFilter
+		var filter_compression *FrontendFilter
+		if len(filters) <= 2 {
+			index_compression_filter := 0
+			index_spoe_filter := 1
+			if filters[0].Type == models.FilterTypeSpoe {
```
Comment (dclaisse) [inline haproxy/state/from_ha.go:62]: What's the goal of this block? Could you please add a comment to explain this inversion?

File: haproxy/state.go [hunk 4]
```diff
@@ -79,6 +79,16 @@ func (h *HAProxy) watch(sd *lib.Shutdown) error {
 				waitAndRetry()
 				continue
 			}
+			// Don't check spoe filter index since it is incremented on dataplane api because retrieved using filters() routine
+			// But requires it to be set to 0 when initializing since filter are created one by one via filter() call
+			for index, current_state_frontend := range currentState.Frontends {
+				if current_state_frontend.FilterSpoe != nil {
+					fromHa.Frontends[index].FilterSpoe.Filter.Index  = current_state_frontend.FilterSpoe.Filter.Index
+				}
```
Comment (dclaisse) [inline haproxy/state.go:87]: Current code uses lowerCamelCase for identifiers, could you please use consistent case?

File: haproxy/state/from_ha.go [hunk 5]
```diff
@@ -51,30 +51,44 @@ func FromHAProxy(ha HAProxyRead) (State, error) {
...
+		if len(logTargets) >= 2 {
+			return state, fmt.Errorf("expected at most 2 filters for frontend %s, got %d", f.Name, len(filters))
 		}
-		var filter *FrontendFilter
-		if len(filters) == 1 {
-			filter = &FrontendFilter{
-				Filter: filters[0],
+		var filter_spoe *FrontendFilter
+		var filter_compression *FrontendFilter
+		if len(filters) <= 2 {
```
Comment (dclaisse) [inline haproxy/state/from_ha.go:59]: What about iterate on filters and check them from their type independently, as you do later (filters[0].Type == models.FilterTypeSpoe), and don't mix logic here? It's hard to understand what condition is for each filter.

File: haproxy/state/from_ha.go [hunk 6]
```diff
@@ -51,30 +51,44 @@ func FromHAProxy(ha HAProxyRead) (State, error) {
...
-			if err != nil {
-				return state, err
+			filter_compression = &FrontendFilter{
+				Filter: filters[index_compression_filter],
 			}
-			if len(binds) != 1 {
-				return state, fmt.Errorf("expected 1 tcp request rule for frontend %s, got %d", f.Name, len(rules))
+
+			if len(filters) == 2 {
+				filter_spoe = &FrontendFilter{
```
Comment (dclaisse) [inline haproxy/state/from_ha.go:71]: How can you be sure SPOE filter is the second one? What about testing its type instead?

File: haproxy/state.go [hunk 7]
```diff
@@ -79,6 +79,16 @@ func (h *HAProxy) watch(sd *lib.Shutdown) error {
 				waitAndRetry()
 				continue
 			}
+			// Don't check spoe filter index since it is incremented on dataplane api because retrieved using filters() routine
```
Comment (dclaisse) [inline haproxy/state.go:82]: Isn't there any way to have a consistent index? Is this behavior expected?

**Review by dclaisse — CHANGES_REQUESTED** [review]

File: haproxy/state/from_ha.go (hunk 2 above)
Comment (mougams) [reply haproxy/state/from_ha.go:55]: Fixed

File: haproxy/state/from_ha.go (hunk 3 above)
Comment (mougams) [reply haproxy/state/from_ha.go]: Simplified

File: haproxy/state.go (hunk 4 above)
Comment (mougams) [reply haproxy/state.go:88]: Done

File: haproxy/state/from_ha.go (hunk 5 above)
Comment (mougams) [reply haproxy/state/from_ha.go:59]: Simplified

File: haproxy/state/from_ha.go (hunk 6 above)
Comment (mougams) [reply haproxy/state/from_ha.go]: Done

File: haproxy/state.go (hunk 7 above)
Comment (mougams) [reply haproxy/state.go]: The API doesn't offer this https://www.haproxy.com/documentation/dataplaneapi/community/#get-/services/haproxy/configuration/filters - you can create only one filter at a time

**Review by dclaisse — CHANGES_REQUESTED** [review]

File: haproxy/state/from_ha.go [hunk 8]
```diff
@@ -51,30 +51,39 @@ func FromHAProxy(ha HAProxyRead) (State, error) {
...
-			if err != nil {
-				return state, err
-			}
-			if len(binds) != 1 {
-				return state, fmt.Errorf("expected 1 tcp request rule for frontend %s, got %d", f.Name, len(rules))
+		var filter_spoe *FrontendFilter
+		var filter_compression *FrontendFilter
+		if len(filters) <= 2 {
+			for _, filter := range filters {
+				if filter.Type == models.FilterTypeSpoe {
```
[the remark this answers is not in the document]
Comment (mougams) [reply haproxy/state/from_ha.go]: Done

**Review by dclaisse — APPROVED** [review]

**Pull request merged**